# Processing FASTA files

Exercise for processing DNA sequences.

* **Contact:** mate.balajti@unibas.ch

In [56]:
import numpy as np

filename1 = "sequences.fasta"
filename2 = "genome.fasta"

ModuleNotFoundError: No module named 'numpy'

## Exercise 1.1 (3 points)

Write a function `parse_fasta()` that takes a path to a FASTA file as input
and returns a tuple of two lists, the first containing sequence headers
stripped of the leading `>`, and the second containing the actual sequences.

> **Notes:**
>  
> * Please write the parser from scratch and do _not_ use existing FASTA
>   parsers, such as the one provided by Biopython.
> * Ensure that wrapped sequences are handled such that fragments of a given
>   sequence are concatenated, without white space, in the order they appear
>   in the file. Make use of the leading `>` character to separate records
>   from each other.
> * Ensure that the number of items in the returned lists correspond to the
>   original number of records in the input file.

In [25]:
def parse_fasta(path: str) -> tuple[list[str], list[str]]:
    
    file = open(path)
    
    header = []
    current_header = ""
    sequence = []
    current_seq = ""

    for line in file:
        line = line.strip() # remove withspaces and newlines
        if line.startswith('>'):
            if current_seq:
                sequence.append(current_seq)
            current_seq = ""
            current_header = line
            header.append(current_header[1:])

        elif line:
            current_seq += line
                
    sequence.append(current_seq)



    return header, sequence

parse_fasta("genome.fasta")

(['chr1', 'chr2', 'chr3', 'chr4'],
 ['TATTTACGCTCCCGCACATATACCGTCGTAAGCGACTGGCTGCTCGCGACCAACCGCCTCGCTACGATTGAGTCGAGAGCGGGTAGACCGGAGGCCGTTCGGCGCATAACTTGTCTTGTTAATTAGGCTGCCTGTGAACGATATTCAATTATGGCACGTGGCGCATTAAGCTCCACGCCACATTCCAACTTCTAGTAGTATGCTGAAACACCGTGCGGGGGACAGAGAGCCATGGCAAATTCAGTTCGGTGTAGACCATACTTTATAACTGGACAACAAGTCCGAGTCCTTCACCTGATGCGCCCAGAACCCCGCATTTCTGTACCTATAGCACCGGCGCACACCCCTCTCACATGCTGGGCAGCAGAGTACTTCTCACTCGATCAGAGAAACCAAACCAGTTGATTGAAGTTCGCATAAGCGTGTTTGCTGTTGCTTGCGGCTCTCATGCACATAGTCGTTCAAGCTTACCCTCCACCGTCGGGCGACAATGCTCGGGGCGCGCATCAGACCGTATTGACTTTTTAGTACAGATTGGACGCAGAGGTTGAGTCATCGTATTCATGTGATAATCTCAACAGGTCCGTCTTGTTATCAAACTTAGTAGTGGGGAGGTGAGTGGTGCAATGTGGAAGTGCCGTAGCCAGCCCGCGATCCACGCACTATTTCAGGGACCGAGGTAACTTAAACCACCCGTTGATGGACATACTGGTTATAGATTTTAGTGCGGGAGTAGGCCTCGTCGCAGCGCAAAGGATGAAGTTTACTAACTGGAGTGGTCAGAAGTTGCCGCCTGTGAAGGCGGCACACGCTTAGTCCGCAATATTATGGGACTGGGGAGGCGACGGTATCCGAAGGTGCTCTGATCAGGATAGCCTCTAGACAGTAATCATGGGTTCTCCATTACGGAGTATTATGGCGGACCGATCGGTCACGTTAATTTGGCTAACTATACGTTCATA

## Exercise 1.2 (2 points)

Write a function `discard_ambiguous_seqs()` that takes two arguments, a list of sequence headers and list of strings as input 
and returns a tuple of two lists with only those headers and strings that exclusively consist of letters of the
"DNA alphabet" (`A`, `C`, `G`, `T`).

> **Notes:**
>  
> * Make sure your implementation is case-insensitive, i.e., sequences
>   containing lowercase DNA characters, even if mixed with uppercase
>   characters, are valid as well.

In [ ]:
def discard_ambiguous_seqs(header: list[str], sequence: list[str]) -> \
  tuple[: list[str], : list[str]]:
    
    r_header = []
    r_sequence = []

    valid_chars = set('ACGT')

    for i in range(len(sequence)):
        seq = sequence[i]

        seq_upper = seq.upper()

        if all (char in valid_chars for char in seq_upper):
          r_header.append(header[i])
          r_sequence.append(seq)
        else:
           print(f"ERROR : {header[i]} has a none valid character -> {seq}")
        
        

    return r_header, r_sequence

list1 = ["chr1", "chr2", "chr3","chr4"]
list2 = ["AAAA", "ACNT", "aTGCgcA", "atcj"]

r_header, r_sequence = discard_ambiguous_seqs(list1, list2)
print("Correct result:")
print("Result headers:", r_header)
print("Result sequences:", r_sequence)

ERROR : chr2 has a none valid character -> ACNT
ERROR : chr4 has a none valid character -> atcj
Correct result:
Result headers: ['chr1', 'chr3']
Result sequences: ['AAAA', 'aTGCgcA']


## Exercise 1.3 (2 points)

Write a function `nucleotide_frequencies()` that takes a list of strings as
input, and which prints out the total frequency of each nucleotide across
all input sequences. Use the following example as a template to format your
output:

```console
A: 0.31
C: 0.21
G: 0.17
T: 0.31
```

> **Notes:**
>
> * Note how numbers are rounded in the example and format decimals printed
>   by your solution in the same manner, i.e., rounded to two significant
>   digits.
> * The function does not require any specific return value. In case you are
>   not aware of how Python deals with functions without an explicit `return`
>   statement, look up the behavior in relevant documentation.

In [54]:
def nucleotide_frequencies(seqs: list[str]) -> None:
    
    total_nucleotides = 0
    count_A = 0
    count_C = 0
    count_T = 0
    count_G = 0

    for seq in seqs:
        for nucleotides in seq:
            if nucleotides == 'A':
                count_A += 1
                total_nucleotides += 1
            if nucleotides == 'C':
                count_C += 1
                total_nucleotides += 1
            if nucleotides == 'T':
                count_T += 1
                total_nucleotides += 1
            if nucleotides == 'G':
                count_G += 1
                total_nucleotides += 1
    
    freq_A = round((count_A / total_nucleotides),2)
    freq_C = round((count_C / total_nucleotides),2)
    freq_T = round((count_T / total_nucleotides),2)
    freq_G = round((count_G / total_nucleotides),2)

    print("A: " + str(freq_A))
    print("C: " + str(freq_C))
    print("T: " + str(freq_T))
    print("G: " + str(freq_G))

test_seq = ["AGCTGGGGAGGCGTTCAGTCTTTCCGAAAAGCGTTTTAAAGCTCCTCCCCTCAACCCTTCGACCTACCTG"]
nucleotide_frequencies(test_seq)

A: 0.2
C: 0.33
T: 0.26
G: 0.21


## Exercise 1.4 (3 points)

Write a function `map_reads()` that takes as input two FASTA files, the first
containing short read sequences ("query"), and the second containing reference
sequences. The function should 
* read the files, 
* discard _query_ sequences that contain non-DNA characters, 
* print the nucleotide fractions for both files to
the console 
* and returns a dictionary of dictionaries, where the outer
dictionary uses the names of query sequences as its keys, and the inner
dictionary uses reference sequence names as keys and a list of 1-based indices
indicating at which position (counting from left to right) in the reference
sequence the query sequence occurs as an exact substring.

Execute the function, passing `sequences.fasta` and `genome.fasta` as input.
Inspect the returned "hits" object (the dictionary of dicionaries). Interpret
the results in at least 2-3 bullet points. What's special about query sequence
`sequence4`?

In [90]:
def map_reads(filename1: str, filename2: str) -> dict[dict[str: list[int]]]:
     
    q_header, q_seq = parse_fasta(filename1)
    r_header, r_seq = parse_fasta(filename2)

    q_header, q_seq = discard_ambiguous_seqs(q_header, q_seq)
    r_header, r_seq = discard_ambiguous_seqs(r_header, r_seq)

    print("Queries :")
    nucleotide_frequencies(q_seq)
    print("Reference :")
    nucleotide_frequencies(r_seq)

    findings = {}
    
    for i, query_seq in enumerate(q_seq):
        if query_seq:  
            query_name = q_header[i]
            findings[query_name] = {}
           
            for j, ref_seq in enumerate(r_seq):
                if ref_seq:  
                    ref_name = r_header[j]
                    
                    positions = []
                    query_len = len(query_seq)
                    ref_len = len(ref_seq)
                    
                
                    for pos in range(ref_len - query_len + 1):
                        if ref_seq[pos:pos + query_len] == query_seq:
                            positions.append(pos + 1)
                    
                    if positions:
                        findings[query_name][ref_name] = positions
    
    return findings
    
map_reads("sequences.fasta", "genome.fasta")


ERROR : sequence3 has a none valid character -> ATGATTDATGTGTCCGGTAACTATAAACGTGCTACGATGT
Queries :
A: 0.23
C: 0.17
T: 0.25
G: 0.36
Reference :
A: 0.25
C: 0.26
T: 0.24
G: 0.25


{'sequence1': {},
 'sequence2': {'chr2': [1422]},
 'sequence4': {'chr2': [1039], 'chr3': [1422], 'chr4': [1455]}}

## Key findings
* sequence1 was not found in any chromosome
* sequence2  was found in chr2 at position 1422
* sequence3 is not valid because htere is a D that is a none valable character.
* sequence4 is shorter than the other so maybe it easier and has a higher probabilty to be found in the chromosomes. Like we can see in the results in appears in chr2 at pos 1039, chr3 at pos 1422, chr4 at pos 1455
